[dvsim] Scheduler updates - max_parallel, max_poll

sriyerg opened the pull request:
The max_parallel setting previously was a Scheduler class variable. It
has now been moved to the Launcher class, since each launcher variant
has different needs.

Likewise, 2 new features are added - `max_poll` and `poll_freq`. By
default, the max_poll 4K, and the polling frequency is set to 1 second.
The purpose of adding these knobs is to reduce the polling rate and the
number of jobs polled at a given time. With external launchers such as
LSF and cloud, polling for a job completion could involve multiple steps
and some external command invocations that are time-consuming. These
features help limit the time taken by the Scheduler to poll the entire
set of jobs. If we do not provide a way to fine-tune / adjust these
parameters, DVSim will appear to have hung (or worse, it could crash the
system).

The max_poll needs to cycle through the list of running jobs, rather
than just the first N ones. For example, if there are 5 jobs running [a,
b, c, d, e], and max_poll is set to 2, on the first poll window, it
polls for jobs a and b to complete. Then it polls b and c, followed by e
and a and so on. To allow this to happen, the list of running jobs now
needs to be maintained as a circular linked list, that supports
iterating through the list, while allowing jobs to be added or removed
from the list in realtime. This is achieved in the newly added
`CircularList` class.

rswarbrick commented on util/dvsim/LauncherFactory.py:
```diff
@@ -43,24 +41,24 @@ def set_launcher_type(is_local=False):
 
+def get_launcher_cls():
+    '''Returns the chosen launcher class.'''
+    global launcher_cls
+    assert launcher_cls is not None
+    return launcher_cls
```
You shouldn't need the `global` here, I don't think. Python's scoping rules are weird, but if a read happens before the first write (IIRC), we'll get the global without needing a hint.

However, could you rename `launcher_cls` to `LAUNCHER_CLS`? The usual Python convention is that module-level globals are capitalized. Also, since this is supposed to be accessed through the getter, maybe it should be `_LAUNCHER_CLS` (to make it clear this is an internal variable, not meant to be exported by the module).

sriyerg replied:
Done.

rswarbrick commented on util/dvsim/Scheduler.py:
```diff
@@ -48,15 +46,23 @@ def __init__(self, items):
         self._failed = {}
         self._killed = {}
         self._total = {}
+        self._targets = CircularList()
+        self._targets.extend(self._scheduled.keys())
```
This is fine, I guess, but it seems rather complicated. If we only need a circular list in one place, wouldn't it be easier just to store an offset and then do the "grab an interval, wrapping around if necessary" logic in a helper method? I'd have thought that would be ~10 lines of code, rather than the 142 in the helper class.

sriyerg replied:
i thought about that as well, but realized that the hard part is grabbing N objects across multiple running targets which would have ended up being a bit more complicated. Plus, we are saving time creating and destroying lists (`to_pass` and `to_fail`) and iterating over those (2 extra loops) every second we poll with this approach (allows our list to be mutated while circularly iterating over it). The current implementation is actually a bit faster especially when running over 1000s of tests.  

IMO a circular linked list is a commonly used data structure. If you look at the implementation, it is fairly easy to follow. Agree that it is a lot of lines, but its mostly to make it resemble a `list` as much as possible. Turns out there is a 3rd party library implementation that I could reuse: https://pypi.org/project/dstructure/
I can update the code to reuse that if you prefer.

rswarbrick replied:
Hmm, ok. I don't think I'd add an external dependency for this (and `dstructure` looks really young with no obvious repository: probably not a great dependency to gain!)

rswarbrick commented on util/dvsim/Scheduler.py:
```diff
@@ -76,6 +82,10 @@ def __init__(self, items):
         # per-target.
         self.item_to_status = {}
 
+        # The chosen launcher class. This allows us to access launcher
+        # variant-specific settings such as max parallel jobs & poll rate.
+        self.launcher_cls = get_launcher_cls()
```
This looks a bit odd to me. Could we pass the chosen launcher class in to the Scheduler constructor? That would move some of the "system construction wiring" logic further up the tree, which I think makes things a bit cleaner.

sriyerg replied:
Done.

rswarbrick commented on util/dvsim/dvsim.py:
```diff
@@ -327,7 +327,8 @@ def parse_args():
                             'Default value 16, unless the DVSIM_MAX_PARALLEL '
                             'environment variable is set, in which case that '
-                            'is used.'))
+                            'is used. Only applicable when launching jobs '
+                            'locally.'))
```
Can we add an error if you pass this argument and we're not launching jobs locally? Silently ignoring arguments causes all manner of confusion to people debugging. If this is absolutely impossible for some reason, at least let's print a warning.

sriyerg replied:
This is doable but a little complicated. You call `resolve_max_args()` in `parse_args()` itself, which sets `args.max_parallel` to 16. We resolve the launcher class later. It is hard to know at that point if `args.max_parallel` was set with `--max-parallel` switch. I would have to save a reference to the original `args.max_parallel` value and check it against the chosen launcher. Let me know if you have a better idea.

rswarbrick replied:
Yeah, you're right: it is a bit difficult. I think it would be an improvement, though. Maybe file an issue to remind us to come back to it? (There are probably other similar cases that would be good to catch).

sriyerg replied:
Done - #6152.

rswarbrick commented on util/dvsim/CircularList.py:
```diff
@@ -0,0 +1,142 @@
+    def __str__(self):
+        return self.item.__str__()
+
+    def __repr__(self):
+        return self.item.__repr__()
```
Is this needed? If not, could you drop it? Hacking `repr` with something that doesn't read as the object you want causes all sorts of problems.

sriyerg replied:
Done.

rswarbrick commented on util/dvsim/CircularList.py:
```diff
@@ -0,0 +1,123 @@
+                    self.tail = previous
+                if self.ptr == current:
+                    self.ptr = previous
+                del current
```
Hmm, what exactly does this line do? I would expect `remove` to unlink things from the data structure. I wouldn't expect it to modify (or call destructors on) the items themselves.

rswarbrick commented on util/dvsim/CircularList.py:
```diff
@@ -0,0 +1,123 @@
+        '''Circularly iterate over the list and return an item.
+
+        This iteration loops back forever, returning an item if available.
+        Returns None if the list is empty.
```
Maybe raise an exception like `StopIteration` if the list is empty? Then we can store an `Optional[X]` in the circular list without any gymnastics.

rswarbrick commented:
Hi Sri! I think we're probably just approaching this from two different standpoints. I favour the "quick bodge" approach of just using an array. I'm kind of surprised that the performance impact of this is measurable, but I have no way of checking for myself. If you're implementing a circular linked list datastructure, I think I'd probably do it slightly differently. But dvsim is your code, so I don't want to stand in your way.

sriyerg commented:
> Hi Sri! I think we're probably just approaching this from two different standpoints. I favour the "quick bodge" approach of just using an array. I'm kind of surprised that the performance impact of this is measurable, but I have no way of checking for myself. If you're implementing a circular linked list datastructure, I think I'd probably do it slightly differently. But dvsim is your code, so I don't want to stand in your way.

Okay. I updated it to do it your way. I got rid of the `CircularList` and added logic to save the last polled index and do the circular looping with that instead.

rswarbrick reviewed (approved):
Looks good to me. The CI failure is unrelated to this PR: a force push will get things green again, but I'd probably just merge.
